core/services/pg/lease_lock.go

```diff
@@ -173,35 +171,27 @@ func (l *leaseLock) logRetry(count int) {
 	}
 }
 
-func (l *leaseLock) Release() {
-	close(l.chStop)
-	l.wg.Wait()
-}
-
-func (l *leaseLock) loop() {
-	defer l.wg.Done()
+func (l *leaseLock) loop(ctx context.Context) {
+	defer close(l.chReleased)
 
 	ticker := time.NewTicker(l.refreshInterval)
 	defer ticker.Stop()
 
-	ctxStop, cancel := utils.ContextFromChan(l.chStop)
-	defer cancel()
-
 	for {
 		select {
-		case <-l.chStop:
+		case <-ctx.Done():
 			ctx, cancel := DefaultQueryCtx()
 			err := multierr.Combine(
 				utils.JustError(l.conn.ExecContext(ctx, `UPDATE lease_lock SET expires_at=NOW() WHERE client_id = $1 AND expires_at > NOW()`, l.id)),
 				l.conn.Close(),
 			)
 			cancel()
 			if err != nil {
-				l.logger.Warnw("Error trying to release lease on shutdown", "err", err)
+				l.logger.Warnw("Error trying to release lease on cancelled ctx", "err", err)
 			}
 			return
 		case <-ticker.C:
-			ctx, cancel := context.WithTimeout(ctxStop, l.leaseDuration)
+			ctx, cancel := context.WithTimeout(context.Background(), l.leaseDuration)
```

Comment:
Why did this one change to `Background()`?

Reply:
Because `ctxStop` has been removed in above and in this particular case we only depend on the timeout. Or do you think it should be the ctx we passed into TakeAndHold?

Reply:
I thought this would still need to yield on shutdown so that a slow call here doesn't block other things, but I suppose in that case the updating the lease to expire now would likely block the same, so maybe that's already unrecoverable.